drivers/virtio/virtio_pci.c

```diff
@@ -0,0 +1,668 @@
+
+	LOG_INF(
+		"virtio pci device with id 0x%x and bdf 0x%x advertised "
+		"feature bits: 0x%.8x%.8x%.8x%.8x",
+		conf->pcie->id,
+		conf->pcie->bdf,
+		virtio_pci_read_device_feature_word(dev, 3),
+		virtio_pci_read_device_feature_word(dev, 2),
+		virtio_pci_read_device_feature_word(dev, 1),
+		virtio_pci_read_device_feature_word(dev, 0)
+	);
+
+	/*
+	 * In case of PCI this should never happen because legacy device would've been caught
+	 * earlier in VIRTIO_PCI_CAP_COMMON_CFG check as this capability shouldn't be present
+	 * in legacy devices, but we are leaving it here as a sanity check
+	 */
+	if (!virtio_pci_read_device_feature_bit(dev, VIRTIO_F_VERSION_1)) {
+		LOG_ERR(
+			"virtio pci device with id 0x%x and bdf 0x%x doesn't advertise "
+			"VIRTIO_F_VERSION_1 feature support",
+			conf->pcie->id,
+			conf->pcie->bdf
+		);
+		return 1;
+	}
+
+	virtio_pci_write_driver_feature_bit(dev, VIRTIO_F_VERSION_1, 1);
+
+	return 0;
+};
+
+struct virtq *virtio_pci_get_virtqueue(const struct device *dev, uint16_t queue_idx)
+{
+	struct virtio_pci_data *data = dev->data;
+
+	return queue_idx < data->virtqueue_count ? &data->virtqueues[queue_idx] : NULL;
+}
+
+void *virtio_pci_get_device_specific_config(const struct device *dev)
+{
+	struct virtio_pci_data *data = dev->data;
+
+	return data->device_specific_cfg;
+}
+
+int virtio_pci_init_virtqueues(const struct device *dev, virtio_enumerate_queues cb, void *opaque)
+{
+	int ret = virtio_pci_set_virtqueues(dev, cb, opaque);
+
+	if (ret != 0) {
+		return ret;
+	}
+	virtio_pci_write_status_bit(dev, DEVICE_STATUS_DRIVER_OK);
+
+	return 0;
+}
+
+int virtio_pci_commit_feature_bits(const struct device *dev)
+{
+	const struct virtio_pci_config *conf = dev->config;
+
+	virtio_pci_write_status_bit(dev, DEVICE_STATUS_FEATURES_OK);
+	if (!virtio_pci_read_status_bit(dev, DEVICE_STATUS_FEATURES_OK)) {
+		LOG_ERR(
+			"virtio pci device with id 0x%x and bdf 0x%x doesn't support selected "
+			"feature bits: 0x%.8x%.8x%.8x%.8x",
+			conf->pcie->id,
+			conf->pcie->bdf,
+			virtio_pci_read_device_feature_word(dev, 3),
+			virtio_pci_read_device_feature_word(dev, 2),
+			virtio_pci_read_device_feature_word(dev, 1),
+			virtio_pci_read_device_feature_word(dev, 0)
+		);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct virtio_driver_api virtio_pci_driver_api = {
+	.get_virtqueue = virtio_pci_get_virtqueue,
+	.notify_queue = virtio_pci_notify_queue,
+	.get_device_specific_config = virtio_pci_get_device_specific_config,
+	.read_device_feature_bit = virtio_pci_read_device_feature_bit,
+	.read_driver_feature_bit = virtio_pci_read_driver_feature_bit,
+	.write_driver_feature_bit = virtio_pci_write_driver_feature_bit_range_check,
+	.commit_feature_bits = virtio_pci_commit_feature_bits,
+	.init_virtqueues = virtio_pci_init_virtqueues
+};
+
+#define VIRTIO_PCI_DEFINE(inst)                                                     \
+	DEVICE_PCIE_INST_DECLARE(inst);                                                 \
```

Comment:
Above this it might be good to ensure that the DT parent has compat `pcie-controller` with a `BUILD_ASSERT()`

Reply:
Done